feature: add txn querying to sdk

CryptoMaxPlanck opened the pull request:
- added query builder to build a gql query
- added query sender to use the gql query and get bridgetxn data back

CryptoMaxPlanck committed:
add:
- query builder to build a gql query
- query sender to use the gql query and get bridgetxn data back

CryptoMaxPlanck commented on src/explorer/gql/utils.ts:
```diff
@@ -0,0 +1,27 @@
+export const BRIDGE_TXN_API_GRAPHQL_URL =
+  "https://syn-explorer-api.metagabbar.xyz/graphql";
+
+export const QUERY_INFO = {
```
better way to do this? like an enum with strings as the value?

bholagabbar replied:
I feel like `QUERY_INFO` is not telling. These are the fields we're returning from the query. So maybe, something like `QUERY_RESPONSE_FIELDS` or something makes more sense

CryptoMaxPlanck commented on src/explorer/send_query.ts:
```diff
@@ -0,0 +1,23 @@
+import { request } from 'graphql-request';
+import { buildTxnQuery } from './gql/query_builder.js';
```
was getting errors when I just did `./gql/query_builder` or `./gql/query_builder.ts`. is this normal to need to have it as a js file even though the native file is ts?

CryptoMaxPlanck commented on src/explorer/gql/utils.ts:
```diff
@@ -0,0 +1,27 @@
+export const BRIDGE_TXN_API_GRAPHQL_URL =
+  "https://syn-explorer-api.metagabbar.xyz/graphql";
+
+export const QUERY_INFO = {
```
also, can add more queries here easily to set specific data types that users want to get returned back

CryptoMaxPlanck commented on src/explorer/gql/utils.ts:
```diff
@@ -0,0 +1,27 @@
+export const BRIDGE_TXN_API_GRAPHQL_URL =
+  "https://syn-explorer-api.metagabbar.xyz/graphql";
```
is this the right URL to use?

bholagabbar replied:
Yes

bholagabbar replied:
Would rename this file to `gqlutils.ts` since we're getting rid of the gql folder

bholagabbar commented on package.json:
```diff
@@ -49,6 +49,7 @@
     "react:remove": "yarn remove --frozen-lockfile react --dev && yarn add --frozen-lockfile --peer react@^17.0.0"
   },
   "dependencies": {
+    "@apollo/client": "^3.3.6",
```
I would get rid of this dependency

bholagabbar commented on package.json:
```diff
@@ -61,6 +62,8 @@
     "@ethersproject/wallet": "^5.5.0",
     "@ethersproject/web": "^5.5.1",
     "ethers": "^5.5.1",
+    "graphql": "^16.5.0",
```
I would get rid of these 2 dependencies as well

bholagabbar commented on src/explorer/gql/query_builder.ts:
```diff
@@ -0,0 +1,78 @@
+import { gql } from 'graphql-request';
```
for the path, I would do `src/explorer/querybuilder.ts` . Don't think you need the gql folder, because you're not defining graphql schemas or types here

bholagabbar replied:
All of this logic is for bridgeTransactions, so more appropriate structure would be `src/explorer/bridgetransactions/query.ts` where you add the logic for everything here.

Don't export the `buildTxnParams` and `buildTxnQuery` while keeping `getBridgeTxnInfo` here in this file itself, exporting it, and importing it in `src/explorer/index.ts`

bholagabbar commented on src/explorer/send_query.ts:
```diff
@@ -0,0 +1,23 @@
+import { request } from 'graphql-request';
```
Don't need this file anymore since. you're moving `getBridgeTxnInfo` into `query.ts`

bholagabbar commented on src/explorer/index.ts:
```diff
@@ -0,0 +1,3 @@
+export { getBridgeTxnInfo } from "./send_query";
+export { buildTxnQuery } from "./gql/query_builder";
+export { BRIDGE_TXN_API_GRAPHQL_URL, QUERY_INFO } from "./gql/utils";
```
only need to export `getBridgeTxnInfo` since this is what the users would really need, the other 2 functions are just for you as helpers

bholagabbar commented on src/explorer/send_query.ts:
```diff
@@ -0,0 +1,23 @@
+  kappa?: string,
+): Promise<JSON> {
+  const query = buildTxnQuery(query_type, chainId, address, txnHash, kappa);
+  return await request(BRIDGE_TXN_API_GRAPHQL_URL, query);
```
What request library is this ? I would use node-fetch ...? sample https://stackoverflow.com/questions/44610310/node-fetch-post-request-using-graphql-query

bholagabbar replied:
Did you test this. Need unit tests

CryptoMaxPlanck committed:
fix:
- remaned files
- fixed exports
- removed specified query results and replaced with one default
- changed graphql querier from graphql-some to native node-fetch

CryptoMaxPlanck committed:
format:
- restructure file format
- move all logic into query.ts

bholagabbar commented on src/explorer/bridge_transactions/query.ts:
```diff
@@ -0,0 +1,99 @@
+  kappa?: string,
+): Promise<JSON> {
+  const query = buildTxnQuery(chainId, address, txnHash, kappa);
+  return (await fetch(BRIDGE_TXN_API_GRAPHQL_URL, {
```
need better exception handling here, log error gracefully

bholagabbar reviewed (changes requested):
unit tests

CryptoMaxPlanck committed:
WIP:
- figuring out testing

CryptoMaxPlanck committed:
add:
- error handling on fetch
- change direct import to fetch from node-fetch

CryptoMaxPlanck committed:
add:
- fully working tests
- named params for the externally facing function
- tests to mocha file

bholagabbar commented on package.json:
```diff
@@ -78,14 +79,15 @@
+    "mocha": "^10.0.0",
     "mocha-lcov-reporter": "1.3.0",
     "mocha-steps": "1.3.0",
+    "react": "^17.0.0",
```
Don't think we need this dependency..? Would remove

bholagabbar commented on test/explorer/explorer-test.ts:
```diff
@@ -0,0 +1,58 @@
+import { QUERY_EXPECTED_OUTPUTS } from "./expected_outputs.js";
+import { expect } from "chai";
+import { getBridgeTxnInfo } from "@sdk/explorer/index.js";
+
```
Would remove the expected output file and dump expected test cases here.

bholagabbar reviewed (commented):
Largely looks good, just address changes and then merge

CryptoMaxPlanck committed:
fix:
- remove expected outputs file and add them directly to test file
- remove react dependency

CryptoMaxPlanck committed:
small

CryptoMaxPlanck committed:
small comment update
